variable/include/scipp/variable/accumulate.h

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file Accumulation functions for variables, based on transform.
+/// @author Simon Heybrock
+#pragma once
+
+#include "scipp/variable/shape.h"
+#include "scipp/variable/transform.h"
+
+namespace scipp::variable {
+
+namespace detail {
+template <class... Ts, class Op, class Var, class... Other>
+static void accumulate(const std::tuple<Ts...> &types, Op op,
+                       const std::string_view &name, Var &&var,
+                       Other &&... other) {
+  // Bail out (no threading) if:
+  // - `other` is implicitly broadcast
+  // - `other` are small, to avoid overhead (important for groupby)
+  // - `other` not const, threading for cumulative ops not possible
+  // - reduction to scalar with more than 1 `other`
+  if ((!other.dims().includes(var.dims()) || ...) ||
+      ((other.dims().volume() < 16384) && ...) || // BM_groupby_large_table
```

Comment:
What exactly i the magic number here?

Reply:
Should we also bail out according to thread limit https://github.com/scipp/scipp/blob/main/CMakeLists.txt#L84?

Reply:
> What exactly i the magic number here?

The element count below which we do not multi thread.

> Should we also bail out according to thread limit

That is handled in `parallel_for`? And even if it was not, I feel it is not worth complicating our code for this debugging feature, given its limited use.

Reply:
How did you come up with the number? In particular, how many threads did you use? The number seems pretty big

Reply:
As the comment indicates, it comes from `BM_groupby_large_table`. I have 12 cores/24 threads.

Reply:
@jl-wynen Note that this is the size of the *input*, if it was the output it would indeed by large.

Reply:
> That is handled in parallel_for? And even if it was not, I feel it is not worth complicating our code for this debugging feature, given its limited use.

Yes, provided that the branching for the new threaded approach does not introduce significant overhead (I cannot see that it does). In any case, I agree, it would be best to avoid complicating the code unnecessarily.